enhancements/ingress/proxy-protocol.md

```diff
@@ -0,0 +1,225 @@
+
+## Motivation
+
+Usually an IngressController is configured with some sort of load balancer in
+front of it.  Typically in such configurations, the connections that the
+IngressController receives have the load balancer's address (rather than the
+original client addresses) as the source address.  However, it is useful to have
+the original source addresses for connections (as received by the load balancer)
+for logging, filtering, and injecting HTTP headers.
+
+Using [PROXY
+protocol](http://www.haproxy.org/download/2.2/doc/proxy-protocol.txt) enables a
+load balancer to preserve the original source addresses and include them when
+forwarding connections to an IngressController, which can then use the original
+source addresses for logging etc.
+
+When OpenShift runs in a cloud platform and an IngressController specifies that
+a service load-balancer should be used, the ingress operator configures the
+service load-balancer and is able to infer based on the specific platform
+whether or not PROXY protocol is needed for preserving source addresses.  For
+example, when an IngressController specifies that it should use an AWS ELB,
+PROXY protocol is required to preserve source addresses, and so the ingress
+operator automatically enables PROXY protocol on the ELB and on the
+IngressController.  Other supported cloud platforms (including Azure and GCP)
+preserve source addresses without the use of PROXY protocol, and so the operator
+does not enable PROXY protocol for service load balancers on these platforms.
+
+On non-cloud platforms, an IngressController is usually configured with an
+external load-balancer, and OpenShift cannot interrogate or infer whether this
+load balancer preserves source addresses or uses PROXY protocol.  Thus cluster
+administrators need to be able to configure whether or not the load balancer and
+IngressController use PROXY protocol.  To this end, this enhancement provides a
+new API that enables cluster administrators to configure PROXY protocol on
+IngressControllers.
+
+This new API is specifically likely to be needed when the IngressController
+specifies the "HostNetwork" or "NodePortService" endpoint publishing strategies,
+which are intended for integrating with external load-balancers.  The API is not
+needed for the "LoadBalancerService" strategy where OpenShift can infer whether
+PROXY protocol is needed, nor is the API needed for the "Private" strategy where
+OpenShift does not publish an endpoint that an external load-balancer could use.
+
+### Goal
+
+1. Enable cluster administrators to enable PROXY protocol on IngressControllers that use the "HostNetwork" and "NodePortService" endpoint publishing strategies.
+
+### Non-Goals
+
+1. Enabling PROXY protocol to be configured on individual Routes.
+1. Enable cluster administrators to enable PROXY protocol on IngressControllers that use the "LoadBalancerService" or "Private" endpoint publishing strategies.
```

Comment:
nit: The non-goals list has 2 items with numbering 1

Reply:
It's a markdown thing.  If each item is numbered with 1, the generator converts it into a properly numbered list.  That's especially nice when dealing with diffs.

Reply:
huh TIL